resource-providers/src/main/java/io/opentelemetry/resource/spring/SpringBootServiceNameGuesser.java

```diff
@@ -0,0 +1,255 @@
+      // expected to fail sometimes
+    }
+    return null;
+  }
+
+  @Nullable
+  private String findByCommandlineArgument() {
+    String result = attemptProcessHandleReflection();
+    if (result == null) {
+      String javaCommand = system.getProperty("sun.java.command");
+      result = parseNameFromCommandLine(javaCommand);
+    }
+    logger.debug("Checking application commandline args: " + result);
+    return result;
+  }
+
+  @Nullable
+  private String attemptProcessHandleReflection() {
+    try {
+      String commandLine = system.attemptGetCommandLineViaReflection();
+      return parseNameFromCommandLine(commandLine);
+    } catch (Exception e) {
+      return null;
+    }
+  }
+
+  private static String parseNameFromCommandLine(String commandLine) {
+    Matcher matcher = COMMANDLINE_PATTERN.matcher(commandLine);
+    boolean ignored = matcher.find(); // Required before group()
+    return matcher.group(1);
+  }
+
+  @Nullable
+  private String readNameFromAppProperties() {
+    return loadFromClasspath(
+        "application.properties", SpringBootServiceNameGuesser::getAppNamePropertyFromStream);
+  }
+
+  @Nullable
+  private static String getAppNamePropertyFromStream(InputStream in) {
+    Properties properties = new Properties();
+    try {
+      properties.load(in);
+      return properties.getProperty("spring.application.name");
+    } catch (IOException e) {
+      return null;
+    }
+  }
+
+  @Nullable
+  private String loadFromClasspath(String filename, Function<InputStream, String> parser) {
+    try (InputStream in = system.openClasspathResource(filename)) {
+      return parser.apply(in);
+    } catch (Exception e) {
+      return null;
+    }
+  }
+
+  // Exists for testing
+  static class SystemHelper {
+
+    String getenv(String name) {
+      return System.getenv(name);
+    }
+
+    String getProperty(String key) {
+      return System.getProperty(key);
+    }
+
+    InputStream openClasspathResource(String filename) {
+      return getClass().getClassLoader().getResourceAsStream(filename);
+    }
+
+    InputStream openFile(String filename) throws Exception {
+      return Files.newInputStream(Paths.get(filename));
+    }
+
+    /**
+     * Attempts to use ProcessHandle to get the full commandline of the current process. Will only
+     * succeed on java 9+.
+     */
+    @SuppressWarnings("unchecked")
+    String attemptGetCommandLineViaReflection() throws Exception {
+      Class<?> clazz = Class.forName("java.lang.ProcessHandle");
+      Method currentMethod = clazz.getDeclaredMethod("current");
+      Method infoMethod = clazz.getDeclaredMethod("info");
+      Object currentInstance = currentMethod.invoke(null);
+      Object info = infoMethod.invoke(currentInstance);
+      Method commandLineMethod = info.getClass().getInterfaces()[0].getMethod("commandLine");
+      Object optionalCommandLine = commandLineMethod.invoke(info);
+      return ((Optional<String>) optionalCommandLine).get();
```

Comment:
Nit: avoid creating NPE:
```suggestion
      return ((Optional<String>) optionalCommandLine).orElse(null);
```